# Data Collection and Understanding

In [1]:
# Import Libraries
import pandas as pd



In [2]:
# load the two data sets

# Algorithm coded crash dataset
algo_df = pd.read_csv("./data/raw/KEN_2012-2023_RTC_v01_M_CSV/ma3route_crashes_algorithmcode.csv")

# Manual coded crash dataset
manual_df = pd.read_csv("./data/raw/KEN_2012-2023_RTC_v01_M_CSV/ma3route_crashes_manualcode.csv")

In [3]:
# View algorithm coded crash dataset

algo_df.head()


,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0


In [4]:
# inspect algorithm crash dataset sizes

algo_df.shape

(31064, 10)

In [5]:
# view manual coded crash dataset

manual_df.head()

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words
0,1,2018-06-06 20:39:54,2018-06-06,-1.263132,36.764283,1,0,0,0,0
1,2,2018-06-30 13:36:54,2018-06-30,-1.295135,36.821932,1,0,0,0,0
2,3,2018-05-26 07:42:54,2018-05-26,-1.371855,36.920457,1,1,0,0,0
3,4,2018-05-26 07:52:24,2018-05-26,-1.209945,36.833230,1,0,0,0,0
4,5,2018-05-26 15:42:24,2018-05-26,-1.215468,36.850638,1,0,0,0,0


In [6]:
# Inspect manual coded crash dataset size

manual_df.shape

(2595, 10)

The algorithm coded crash dataset has 31064 accident records/observations and 10 varibales or features while the manual coded crash dataset has 2595 accident records/observations and 10 features or variables.

In [7]:
# Verifying the columns of algorithm coded crash dataset

for features in algo_df.columns:
    print(features)


crash_id
crash_datetime
crash_date
latitude
longitude
n_crash_reports
contains_fatality_words
contains_pedestrian_words
contains_matatu_words
contains_motorcycle_words


In [8]:
# Verifying the columns of manual coded crash dataset

for variables in manual_df:
    print(variables)

crash_id
crash_datetime
crash_date
latitude
longitude
n_crash_reports
contains_fatality_words
contains_pedestrian_words
contains_matatu_words
contains_motorcycle_words


Both datasets (algorithm and manual coded datasets) have the similar columns/features/varibales. 

In [9]:
# combine the Datasets

crashes_df = pd.concat([algo_df, manual_df], ignore_index = True)
crashes_df

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0
...,...,...,...,...,...,...,...,...,...,...
33654,3111,2018-05-23 07:23:54,2018-05-23,-1.434153,36.965546,4,0,0,1,0
33655,3112,2018-05-23 14:38:24,2018-05-23,-1.329874,36.869561,3,0,0,0,0
33656,3113,2018-05-24 07:45:24,2018-05-24,-1.269935,36.792622,18,1,0,1,0
33657,3114,2018-05-24 08:12:24,2018-05-24,-1.325156,36.848661,2,0,0,0,0


In [10]:
# Checking the Combined Dataset
crashes_df.shape


(33659, 10)

We now have a total of 33,659 accident records and 10 variables/features

In [11]:
# view the first records

crashes_df.head()

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0


In [12]:
# dataset information

crashes_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 33659 entries, 0 to 33658
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   crash_id                   33659 non-null  int64  
 1   crash_datetime             33659 non-null  str    
 2   crash_date                 33659 non-null  str    
 3   latitude                   33659 non-null  float64
 4   longitude                  33659 non-null  float64
 5   n_crash_reports            33659 non-null  int64  
 6   contains_fatality_words    33659 non-null  int64  
 7   contains_pedestrian_words  33659 non-null  int64  
 8   contains_matatu_words      33659 non-null  int64  
 9   contains_motorcycle_words  33659 non-null  int64  
dtypes: float64(2), int64(6), str(2)
memory usage: 2.6 MB


In [13]:
crashes_df.describe()

,crash_id,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words
count,33659.000000,33659.000000,33659.000000,33659.000000,33659.000000,33659.000000,33659.000000,33659.000000
mean,14456.273538,-1.272371,36.853191,1.415015,0.072789,0.029858,0.081702,0.036751
std,9388.643947,0.138422,0.138351,1.579152,0.259793,0.170199,0.273914,0.188152
min,1.000000,-4.059868,34.145808,1.000000,0.000000,0.000000,0.000000,0.000000
25%,5820.500000,-1.316585,36.799301,1.000000,0.000000,0.000000,0.000000,0.000000
50%,14235.000000,-1.280129,36.842001,1.000000,0.000000,0.000000,0.000000,0.000000
75%,22649.500000,-1.244702,36.891573,1.000000,0.000000,0.000000,0.000000,0.000000
max,31064.000000,1.257331,40.171389,67.000000,1.000000,1.000000,1.000000,1.000000


In [14]:
# missing values
crashes_df.isnull().sum()

crash_id                     0
crash_datetime               0
crash_date                   0
latitude                     0
longitude                    0
n_crash_reports              0
contains_fatality_words      0
contains_pedestrian_words    0
contains_matatu_words        0
contains_motorcycle_words    0
dtype: int64

There are no missing records in the combined dataset

In [15]:
# check for duplicates

crashes_df.duplicated().sum()

np.int64(0)

There are no duplicated records

In [16]:
# save the combined dataset

crashes_df.to_csv("./data/processes/Combined algorithm and manual crash datasets", index =False
)

## Dataset Overview

The dataset was created by combining:

1. ma3route_crashes_algorithmcode.csv
2. ma3route_crashes_manualcode.csv

The combined dataset contains road traffic crash records collected between 2012 and 2023.

Key variables include:
- Date and time of crash
- Geographic coordinates
- Fatality indicators
- Pedestrian involvement
- Motorcycle involvement
- Matatu involvement

There are no missing records nor duplicated records, therefore we have a total of 33,659 accident records and 10 variables.

# Data Cleaning and Preparation

In [17]:
# create a working copy
df_accidents = crashes_df.copy()

I have done this to avoid messing up or damaging the original dataset i.e crashes_df

In [18]:
# check the data types

df_accidents.dtypes

crash_id                       int64
crash_datetime                   str
crash_date                       str
latitude                     float64
longitude                    float64
n_crash_reports                int64
contains_fatality_words        int64
contains_pedestrian_words      int64
contains_matatu_words          int64
contains_motorcycle_words      int64
dtype: object

crash_datetime and crash_date columns have to be converted to a date format from text or string format so that it can be easier to extract
- Year
- Month
- Day
- Hour

This will help us answer questions in our objectives such as what year recorded the highest number of crashes, Identify high risk months many other questions.

In [19]:
# Convert date columns
df_accidents['crash_datetime'] = pd.to_datetime(df_accidents['crash_datetime'])

df_accidents['crash_date'] = pd.to_datetime(df_accidents['crash_date'])

In [20]:
# verify the conversions
df_accidents[['crash_datetime', 'crash_date']].dtypes

crash_datetime    datetime64[us]
crash_date        datetime64[us]
dtype: object

In [21]:
# create year column
df_accidents['year'] = df_accidents['crash_datetime'].dt.year

# create month column
df_accidents['month'] = df_accidents['crash_datetime'].dt.month

df_accidents

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0,2018,6
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0,2018,8
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0,2018,5
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0,2018,5
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0,2018,5
...,...,...,...,...,...,...,...,...,...,...,...,...
33654,3111,2018-05-23 07:23:54,2018-05-23,-1.434153,36.965546,4,0,0,1,0,2018,5
33655,3112,2018-05-23 14:38:24,2018-05-23,-1.329874,36.869561,3,0,0,0,0,2018,5
33656,3113,2018-05-24 07:45:24,2018-05-24,-1.269935,36.792622,18,1,0,1,0,2018,5
33657,3114,2018-05-24 08:12:24,2018-05-24,-1.325156,36.848661,2,0,0,0,0,2018,5


In [22]:
# create the month name

df_accidents['month_name'] = df_accidents['crash_datetime'].dt.month_name()
df_accidents

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month,month_name
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0,2018,6,June
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0,2018,8,August
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0,2018,5,May
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0,2018,5,May
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0,2018,5,May
...,...,...,...,...,...,...,...,...,...,...,...,...,...
33654,3111,2018-05-23 07:23:54,2018-05-23,-1.434153,36.965546,4,0,0,1,0,2018,5,May
33655,3112,2018-05-23 14:38:24,2018-05-23,-1.329874,36.869561,3,0,0,0,0,2018,5,May
33656,3113,2018-05-24 07:45:24,2018-05-24,-1.269935,36.792622,18,1,0,1,0,2018,5,May
33657,3114,2018-05-24 08:12:24,2018-05-24,-1.325156,36.848661,2,0,0,0,0,2018,5,May


The months have been named as either January, February and so on rather than 1st, 2nd.

In [23]:
# create day of the week
df_accidents['day_name'] = (df_accidents['crash_datetime'].dt.day_name())

df_accidents

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month,month_name,day_name
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0,2018,6,June,Wednesday
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0,2018,8,August,Friday
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0,2018,5,May,Friday
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0,2018,5,May,Friday
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0,2018,5,May,Friday
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
33654,3111,2018-05-23 07:23:54,2018-05-23,-1.434153,36.965546,4,0,0,1,0,2018,5,May,Wednesday
33655,3112,2018-05-23 14:38:24,2018-05-23,-1.329874,36.869561,3,0,0,0,0,2018,5,May,Wednesday
33656,3113,2018-05-24 07:45:24,2018-05-24,-1.269935,36.792622,18,1,0,1,0,2018,5,May,Thursday
33657,3114,2018-05-24 08:12:24,2018-05-24,-1.325156,36.848661,2,0,0,0,0,2018,5,May,Thursday


This will help in knowing what days had the most accidents. 

In [24]:
# create hour
df_accidents['hour'] = df_accidents['crash_datetime'].dt.hour

df_accidents.head()

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month,month_name,day_name,hour
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0,2018,6,June,Wednesday,20
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0,2018,8,August,Friday,6
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0,2018,5,May,Friday,17
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0,2018,5,May,Friday,18
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0,2018,5,May,Friday,21


These variables will assist in identifying morning rush hour, afternoon traffic, evening congestion and latenight accidents

In [25]:
# check missing values again
df_accidents.isnull().sum()

crash_id                     0
crash_datetime               0
crash_date                   0
latitude                     0
longitude                    0
n_crash_reports              0
contains_fatality_words      0
contains_pedestrian_words    0
contains_matatu_words        0
contains_motorcycle_words    0
year                         0
month                        0
month_name                   0
day_name                     0
hour                         0
dtype: int64

All records were perfectly converted

In [26]:
# inspect missing coordinates
df_accidents[df_accidents['latitude'].isnull() |
    df_accidents['longitude'].isnull()
    ]

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month,month_name,day_name,hour


Coordinates will assist in hotspot analysis that's why it is important to know how many are missing and if the number is significant 

### check uniques values in crash indicators

In [27]:
df_accidents['contains_fatality_words'].value_counts(dropna=False)

contains_fatality_words
0    31209
1     2450
Name: count, dtype: int64

In [28]:
df_accidents['contains_pedestrian_words'].value_counts(dropna=False)

contains_pedestrian_words
0    32654
1     1005
Name: count, dtype: int64

In [29]:
df_accidents['contains_motorcycle_words'].value_counts(dropna=False)

contains_motorcycle_words
0    32422
1     1237
Name: count, dtype: int64

In [30]:
df_accidents['contains_matatu_words'].value_counts(dropna=False)

contains_matatu_words
0    30909
1     2750
Name: count, dtype: int64

Checking unique values help determine wheter further cleaning is needed. 

### Save the new and cleaned dataset

In [31]:
#This will become the new dataset to be used in our future notebooks
df_accidents.to_csv(
    "./data/processes/Kenya_road_accidents_cleaned.csv", 
    index = False
)


In [32]:
df_accidents.head()

,crash_id,crash_datetime,crash_date,latitude,longitude,n_crash_reports,contains_fatality_words,contains_pedestrian_words,contains_matatu_words,contains_motorcycle_words,year,month,month_name,day_name,hour
0,1,2018-06-06 20:39:54,2018-06-06,-1.263030,36.764374,1,0,0,0,0,2018,6,June,Wednesday,20
1,2,2018-08-17 06:15:54,2018-08-17,-0.829710,37.037820,1,1,0,0,0,2018,8,August,Friday,6
2,3,2018-05-25 17:51:54,2018-05-25,-1.125301,37.003297,1,0,0,0,0,2018,5,May,Friday,17
3,4,2018-05-25 18:11:54,2018-05-25,-1.740958,37.129026,1,0,0,0,0,2018,5,May,Friday,18
4,5,2018-05-25 21:59:54,2018-05-25,-1.259392,36.842321,1,1,0,0,0,2018,5,May,Friday,21


## Data Cleaning Summary

The following cleaning steps were performed:

1. Combined algorithm-coded and manually-coded datasets.
2. Removed duplicate records.
3. Converted date columns to datetime format.
4. Created new variables:
   - year
   - month
   - month_name
   - day_name
   - hour
5. Inspected missing values and coordinate completeness.
6. Saved a cleaned version of the dataset for analysis.